internal/auth/providers/okta.go

```diff
@@ -0,0 +1,416 @@
+	params.Add("code", code)
+	params.Add("grant_type", "authorization_code")
+	params.Add("scope", p.Scope)
+
+	var response struct {
+		AccessToken  string `json:"access_token"`
+		RefreshToken string `json:"refresh_token"`
+		ExpiresIn    int64  `json:"expires_in"`
+		IDToken      string `json:"id_token"`
+	}
+	err := p.oktaRequest("POST", p.RedeemURL.String(), params, []string{"action", "redeem"}, nil, &response)
+	if err != nil {
+		return nil, err
+	}
+	var email string
+	email, err = p.verifyEmailWithAccessToken(response.AccessToken)
+	if err != nil {
+		return nil, err
+	}
+	return &sessions.SessionState{
+		AccessToken:  response.AccessToken,
+		RefreshToken: response.RefreshToken,
+
+		RefreshDeadline:  sessions.ExtendDeadline(time.Duration(response.ExpiresIn) * time.Second),
+		LifetimeDeadline: sessions.ExtendDeadline(p.SessionLifetimeTTL),
+		Email:            email,
+	}, nil
+}
+
+// verifyEmailWithAccessToken takes in an access token and
+// checks if it is verified with Okta. Conditionally returns
+// an empty string, or if validated the email address in question.
+func (p *OktaProvider) verifyEmailWithAccessToken(AccessToken string) (string, error) {
+	if AccessToken == "" {
+		return "", ErrBadRequest
+	}
+
+	userinfo, err := p.UserInfo(AccessToken)
+	if err != nil {
+		return "", err
+	}
+	if userinfo.EmailAddress == "" {
+		return "", errors.New("missing email")
+	}
+	if !userinfo.EmailVerified {
+		return "", fmt.Errorf("email %s not listed as verified", userinfo.EmailAddress)
+	}
+
+	return userinfo.EmailAddress, nil
+}
+
+// ValidateGroupMembership takes in an email, a list of allowed groups an access token
+// - sends a request to the /userinfo endpoint to return a slice of users' group membership
+// - compares allowed groups to users' groups.
+// Conditionally returns an empty (nil) slice or a slice of the matching groups.
+func (p *OktaProvider) ValidateGroupMembership(email string, allowedGroups []string, accessToken string) ([]string, error) {
+	if accessToken == "" {
+		return nil, ErrBadRequest
+	}
+
+	if len(allowedGroups) == 0 {
+		return []string{}, nil
+	}
+
+	userinfo, err := p.UserInfo(accessToken)
+	if err != nil {
+		return nil, err
+	}
+	if len(userinfo.Groups) == 0 {
+		return nil, fmt.Errorf("no group membership returned for %s", email)
+	}
+
+	matchingGroups := []string{}
+	for _, x := range allowedGroups {
+		for _, y := range userinfo.Groups {
+			if x == y {
+				matchingGroups = append(matchingGroups, x)
+			}
+		}
+	}
+	return matchingGroups, nil
+}
+
+// UserInfo takes in an access token and sends a request to the /userinfo endpoint.
+// Conditionally returns nil response or a struct of specified claims.
+func (p *OktaProvider) UserInfo(AccessToken string) (*UserInfoResponse, error) {
+	// https://developer.okta.com/docs/api/resources/oidc/#userinfo
+	response := &UserInfoResponse{}
+	if AccessToken == "" {
+		return nil, ErrBadRequest
+	}
+	var bearer = "Bearer " + AccessToken
```

Comment:
let's string format rather than string concat

```golang
bearer := fmt.Sprintf("Bearer %s", AccessToken)
```

I also typically prefer `:=` variable instantiation over `var x`.